test/core/event_engine/posix/event_poller_posix_test.cc

```diff
@@ -513,7 +544,196 @@ TEST_P(EventPollerTest, TestEventPollerHandleChange) {
+                  Unref();
+                }));
+              }
+            })) {
+    ++kTotalActiveWakeupFdHandles;
+    EXPECT_GT(num_wakeups_, 0);
+    EXPECT_NE(scheduler_, nullptr);
+    EXPECT_NE(poller_, nullptr);
+    wakeup_fd_ = *PipeWakeupFd::CreatePipeWakeupFd();
+    handle_ = poller_->CreateHandle(wakeup_fd_->ReadFd(), "test", false);
+    EXPECT_NE(handle_, nullptr);
+    handle_->NotifyOnRead(on_read_);
+    //  Send a wakeup initially.
+    EXPECT_TRUE(wakeup_fd_->Wakeup().ok());
+  }
+
+  ~WakeupFdHandle() { delete on_read_; }
+
+ private:
+  absl::Status ReadPipe() {
+    char buf[128];
+    ssize_t r;
+    int total_bytes_read = 0;
+    for (;;) {
+      r = read(wakeup_fd_->ReadFd(), buf, sizeof(buf));
+      if (r > 0) {
+        total_bytes_read += r;
+        continue;
+      }
+      if (r == 0) return absl::OkStatus();
+      switch (errno) {
+        case EAGAIN:
+          return total_bytes_read > 0 ? absl::OkStatus()
+                                      : absl::InternalError("Spurious Wakeup");
+        case EINTR:
+          continue;
+        default:
+          return absl::Status(absl::StatusCode::kInternal,
+                              absl::StrCat("read: ", strerror(errno)));
+      }
+    }
+  }
+  void Ref() { ref_count_.fetch_add(1, std::memory_order_relaxed); }
+  void Unref() {
+    if (ref_count_.fetch_sub(1, std::memory_order_acq_rel) == 1) {
+      // Once the handle has orphaned itself, decrement
+      // kTotalActiveWakeupFdHandles. Once all handles have orphaned themselves,
+      // send a Kick to the poller.
+      handle_->OrphanHandle(
+          PosixEngineClosure::TestOnlyToClosure(
+              [poller = poller_, wakeupfd_handle = this](absl::Status status) {
+                EXPECT_TRUE(status.ok());
+                if (--kTotalActiveWakeupFdHandles == 0) {
+                  poller->Kick();
+                }
+                delete wakeupfd_handle;
+              }),
+          nullptr, "");
+    }
+  }
+  int num_wakeups_;
+  Scheduler* scheduler_;
+  PosixEventPoller* poller_;
+  PosixEngineClosure* on_read_;
+  std::atomic<int> ref_count_{1};
+  std::unique_ptr<WakeupFd> wakeup_fd_;
+  EventHandle* handle_;
+};
+
+// A helper class to create Fds and drive the polling for these Fds. It
+// repeatedly calls the Work(..) method on the poller to get pet pending events,
+// then schedules another parallel Work(..) instantiation and processes these
+// pending events. This continues until all Fds have orphaned themselves.
+class Worker {
+ public:
+  Worker(Scheduler* scheduler, PosixEventPoller* poller, int num_handles,
+         int num_wakeups_per_handle)
+      : scheduler_(scheduler), poller_(poller) {
+    handles_.reserve(num_handles);
+    for (int i = 0; i < num_handles; i++) {
+      handles_.push_back(
+          new WakeupFdHandle(num_wakeups_per_handle, scheduler_, poller_));
+    }
+  }
+  void Ref() { ref_count_.fetch_add(1, std::memory_order_relaxed); }
```

Comment:
nit: You can use our internal RefCounted types here, no need to reinvent.

Reply:
Done. I used DualRefCounted now.